014-quarkus-panache-with-transactions-xa/src/test/java/io/quarkus/qe/dbpool/AgroalPoolTest.java

```diff
@@ -0,0 +1,188 @@
+import io.smallrye.mutiny.Multi;
+import io.smallrye.mutiny.tuples.Tuple2;
+
+@QuarkusTest
+@TestProfile(MysqlTestProfile.class)
+@TestMethodOrder(OrderAnnotation.class)
+public class AgroalPoolTest {
+
+    private static final Logger LOG = Logger.getLogger(AgroalPoolTest.class);
+
+    @Inject
+    EntityManager em;
+
+    @Inject
+    UserRepository users;
+
+    @Inject
+    AgroalDataSource agroalDataSource;
+
+    static final int TIMEOUT_SEC = 60;
+
+    @ConfigProperty(name = "quarkus.datasource.jdbc.idle-removal-interval")
+    String idleSec;
+
+    @ConfigProperty(name = "quarkus.datasource.jdbc.max-size")
+    int datasourceMaxSize;
+
+    @ConfigProperty(name = "quarkus.datasource.jdbc.min-size")
+    int datasourceMinSize;
+
+    @Test
+    @Order(1)
+    public void idleTimeoutTest() throws InterruptedException {
+        makeApplicationQuery();
+        Thread.sleep(3000);
+        assertEquals(1, activeConnections(), "agroalCheckIdleTimeout: Expected " + datasourceMinSize + " active connections");
+    }
+
+    @Test
+    @Order(2)
+    public void poolTurnoverTest() throws InterruptedException {
+        final int events = 2500;
+        final AtomicInteger max = new AtomicInteger(0);
+        CountDownLatch queuedEvents = new CountDownLatch(events);
+
+        Multi<Tuple2<Long, Long>> combined = Multi.createBy().combining()
+                .streams(makeApplicationQueryAsync(events), activeConnectionsAsync(events)).asTuple();
+
+        combined.subscribe().with(t -> {
+            Long activeCon = t.getItem2();
+            if (max.intValue() < activeCon) {
+                max.set(activeCon.intValue());
+            }
+            assertTrue(datasourceMaxSize >= activeCon, "More mysql SQL connections than pool max-size");
+            assertTrue(datasourceMinSize <= activeCon, "Less mysql SQL connections than pool min-size");
+            queuedEvents.countDown();
+        });
+
+        LOG.info("agroalCheckDbPoolTurnover: Max connection requested: " + max);
+        queuedEvents.await(TIMEOUT_SEC, TimeUnit.SECONDS);
+        assertEquals(0, queuedEvents.getCount(),
+                String.format("agroalCheckDbPoolTurnover: Missing %d events.", queuedEvents.getCount()));
+    }
+
+    @RepeatedTest(10)
+    @Order(3)
+    public void borderConditionBetweenIdleAndGetConnectionTest() throws InterruptedException {
+        final int concurrentRequestAmount = 2;
+        CountDownLatch queuedEvents = new CountDownLatch(concurrentRequestAmount);
+        Multi.createFrom().range(0, concurrentRequestAmount).subscribe().with(n -> Multi.createFrom().ticks()
+                .every(Duration.ofSeconds(getIdle() + 3))
+                .onOverflow().drop()
+                .onFailure().invoke(e -> Assertions.fail("Unexpected exception " + e.getMessage()))
+                .subscribe().with(i -> {
+                    makeApplicationQuery();
+                    queuedEvents.countDown();
+                }));
```

Comment:
I don't understand very well this code.
Are you running two concurrent threads and each thread is making calls to the application every 3 seconds? Then, we verify that the queuedEvents is 0... 
Why do we need 2 threads? If one of them fails, the another thread can call the application twice and then the test will still satisfy the condition.

Reply:
is not every 3 sec. We are making a query 3 ms ahead of the IDLE-timeout. 

`.every(Duration.ofSeconds(getIdle() + 3))`

We are checking border condition errors.

Reply:
I still don't understand this method. As I can read: `.every(Duration.ofSeconds(1 + 3))`, It's traslated to every 4 seconds. If it's for some reason, it's indeed 3 or 4 ms, can you make it more readable please?

Reply:
Ok I have changed a little bit the signature in order to be more readable:

`.every(Duration.ofSeconds(getIdleMs() + 3))`

- 2 concurrent threads and then making a request every X time: 

concurrent queries every X secs N times (Repeat annotation) is exactly the aim of this test. This is why is called `connectionConcurrencyTest`, or in other words, check that two (or more) concurrent queries completed successfully. In case of any issue, ` queuedEvents.countDown()` is not going to be invoked and the test is going to fail 

```
assertEquals(0, queuedEvents.getCount(),
                String.format("agroalConnectionConcurrency: Missing %d events.", queuedEvents.getCount()));
```

 - If what we want is to stress the db pool, what about running this test using more than 2 threads (which I think it simulates the users):
 
  we could increase the number of concurrent queries (I did , is more expensive in terms of CPU and time, I have increased from 2 to 20) but, the `stress` feature is cover by the `RepeatedTest` annotation (is a combination of both). So, 2 concurrent queries repeated 500 times, means 1000 consecutive successful queries. 

With the current concurrent configuration, attending to your request, we are going to make 10k consecutive successful queries.  Is going to be more expensive(10 times), I thought that 1000 queries were enough to probe that we handler concurrency as expected ( some issues have risen with even less iteration and concurrency), from a agroal datasource point of view, but is also ok)

Reply:
I see you have changed the every method to:
`.every(Duration.ofSeconds(getIdleMs() + 3))`

But it's still odd to me as you have milliseconds and use it as seconds.
Let me be clearer, this is what you are doing:

```
int concurrentRequestAmount = 2; // To mimic users
int idle = 1; // 1 seconds
long idleMs = Duration.ofSeconds(idle).toMillis(); // 1000 milliseconds
CountDownLatch queuedEvents = new CountDownLatch(concurrentRequestAmount); // Wait for 2 requests
Multi.createFrom().range(0, concurrentRequestAmount).subscribe()
        .with(n -> Multi.createFrom().ticks()
                .every(Duration.ofSeconds(idleMs + 3)) // 1000 + 3 as seconds 
                .onOverflow().drop()
                .subscribe().with(i -> {
                    System.out.println("Tick: " + i);
                    queuedEvents.countDown();
                }));
assertTrue(queuedEvents.await(5, TimeUnit.SECONDS));     
```

This is working for you, because you're waiting for 2 requests done by 2 concurrent requests, but the `Multi.createFrom().ticks().every(Duration.ofSeconds(idleMs + 3))` is doing a request every 1003 seconds now (before it was 4 seconds).

I think the right code is to change it to `Multi.createFrom().ticks().every(Duration.ofMillis(idleMs + 3))`.

However, as you are verifying that the requests of the concurrent users worked fine, the current solution would be the same as doing:

```
final int concurrentRequestAmount = 20;
CountDownLatch queuedEvents = new CountDownLatch(concurrentRequestAmount);
Multi.createFrom().range(0, concurrentRequestAmount).subscribe().with(n -> {
            makeApplicationQuery();
            queuedEvents.countDown();
        });
```

But guessing that what you exactly want to stress the border condition, then another idea would be to either assert there are no failures after some amount of time or verify the requests after some ticks (`new CountDownLatch(concurrentRequestAmount * NUM_OF_ITERATIONS or ticks);`

Reply:
Yes in this test `borderConditionBetweenIdleAndGetConnectionTest` the key concept is the IDLE-timeout border condition.